### PR TITLE
Windows 1809 Support Alpha

### PR DESCRIPTION
**Pull request by thxCode**
- [x] Support Windows core 1809
- [x] Bug fix:
  - [x]  **`nginx-proxy` container sometimes fails to start**: https://github.com/rancher/rancher/issues/16074
  - [x] **kubelet (Kubernetes 1.12.) wont start on windows**: https://github.com/rancher/rancher/issues/17341
  - [x] **Node scheduling problem Linux/Windows for Rancher System components**: https://github.com/rancher/rancher/issues/17423
  - [x] **Support upgrade**: https://github.com/rancher/rancher/issues/19048
  - [x] **Windows Nodes crashing after 2-3 days**: https://github.com/rancher/rancher/issues/17499
- [x] Refactor Windows logic
- [x] VxLan support

**Issues:**
- https://github.com/rancher/rancher/issues/16460 (Track)
- https://github.com/rancher/rancher/issues/19173

**Related PR:**
- https://github.com/rancher/types/pull/750
- https://github.com/rancher/kubernetes-windows-dockerfiles/pull/7
- https://github.com/rancher/rke/pull/1290

### CODE REVIEW

File: pkg/rkeworker/execute_linux.go [hunk 1]
```diff
@@ -0,0 +1,28 @@
+package rkeworker
+
+import (
+	"context"
+	"strings"
+)
+
+func init() {
```
Comment (orangedeng) [inline pkg/rkeworker/execute_linux.go]: Don't need it.

File: pkg/rkeworker/execute_windows.go [hunk 2]
```diff
@@ -1,106 +1,69 @@
...
-	if writeCertOnly {
-		return nil
-	}
-
-	// pre run docker
-	for name, process := range nodeConfig.Processes {
-		if strings.HasPrefix(name, "pre-run-docker") {
-			if err := runProcess(ctx, process.Name, process, true, false); err != nil {
-				return err
+func init() {
```
Comment (orangedeng) [inline pkg/rkeworker/execute_windows.go]: don't need it

File: pkg/rkeworker/execute_windows.go [hunk 3]
```diff
@@ -1,106 +1,69 @@
...
-		return executePowerShell(ctx, scriptPath, hyperkubeProcess.Args...)
+		return nil
 	}
 }
 
 func Stop(ctx context.Context) error {
-	scriptPath := rancherDir + `\stop.ps1`
-	if _, err := os.Stat(scriptPath); err != nil {
-		return fmt.Errorf(`can't find %s while stopping`, scriptPath)
+	scriptPath, err := getPowerShellScript("stop")
```
Comment (orangedeng) [inline pkg/rkeworker/execute_windows.go]: You are following the same pattern.
```
1. getPowerShellScript
2. executePowerShell
```
Maybe we can combine both function together.

File: pkg/rkeworker/execute_windows.go [hunk 4]
```diff
@@ -167,62 +130,32 @@ func executePowerShell(rootCtx context.Context, scriptPath string, args ...strin
 	g.Go(func() error {
 		if err := command.Run(); err != nil {
 			if exitError, ok := err.(*exec.ExitError); ok {
-				waitStatus := exitError.Sys().(syscall.WaitStatus)
-				return powerShellExistStatusError(waitStatus.ExitStatus())
+				switch exitError.ExitCode() {
+				case 0:
+					return nil
+				case 2:
```
Comment (orangedeng) [inline pkg/rkeworker/execute_windows.go]: do we missing `case 1:` for `ErrHyperKubePSScriptCrash `.

File: scripts/windows/build-agent.ps1 [hunk 5]
```diff
@@ -11,7 +11,7 @@ $imageTag = "rancher/rancher-agent:$version"
 
 pushd $rootPath
 Write-Host "Running agent build"
-if ($version.endswith("-1803")) {
```
Comment (orangedeng) [inline scripts/windows/build-agent.ps1]: I think we need to support building agent for both `1803` and `1809`

File: package/windows/hyperkube.ps1 [hunk 6]
```diff
@@ -516,33 +603,25 @@ function start-flanneld {
...
+    $fgRun = get-env-var -Key "CATTLE_AGENT_FG_RUN"
     $flanneldArgs = @(
         "`"--kubeconfig-file=$KubeDir\ssl\kubecfg-kube-node.yaml`""
         "`"--iface=$NodeIP`""
         "`"--ip-masq`""
         "`"--kube-subnet-mgr`""
         "`"--iptables-forward-rules=false`""
+        "`"--logtostderr=$fgRun`""
+        "`"--alsologtostderr=true`""
+        "`"--log-file=$LogDir\flanneld.log`""
```
[the remark this answers is not in the document]
Comment (orangedeng) [reply package/windows/hyperkube.ps1]: I see it now. The custom flannel binary is maintained in https://github.com/thxCode/coreos-flannel

File: package/windows/Dockerfile.agent [hunk 7]
```diff
@@ -1,6 +1,6 @@
 ARG PS_VERSION
 
-FROM golang:1.11.0-windowsservercore-1803 AS build-stage
+FROM golang:1.12.1-windowsservercore-1809 AS build-stage
```
[the remark this answers is not in the document]
Comment (thxCode) [reply package/windows/Dockerfile.agent]: Can I patch this with a new PR to support Windows Docker manifest and Windows CI? On the another hand, `1809` host can package `1803` container with Hyper-V mode.

File: package/windows/hyperkube.ps1 (hunk 6 above)
[the remark this answers is not in the document]
Comment (thxCode) [reply package/windows/hyperkube.ps1]: Yes, https://github.com/thxCode/coreos-flannel/commits/rancher_package

File: pkg/agent/main_windows.go [hunk 8]
```diff
@@ -226,50 +95,56 @@ func getTokenAndURL() (string, string, error) {
 }
 
 func isConnected() bool {
-	_, err := os.Stat(connectedPath)
+	_, err := os.Stat("connected")
```
[the remark this answers is not in the document]
Comment (thxCode) [reply pkg/agent/main_windows.go]: Yes, this part is similar to the Linux agent.

File: pkg/agent/windows/remotedialer/client_windows.go [hunk 9]
```diff
@@ -69,20 +72,24 @@ func connectToProxyWhileWindows(rootContext context.Context, proxyURL string, he
 				return err
 			}()
 			if err != nil {
```
[the remark this answers is not in the document]
Comment (thxCode) [reply pkg/agent/windows/remotedialer/client_windows.go]: updated

File: pkg/agent/windows/remotedialer/client_windows.go [hunk 10]
```diff
@@ -53,8 +57,7 @@ func connectToProxyWhileWindows(rootContext context.Context, proxyURL string, he
 	}
 
 	eg.Go(func() error {
-		reconnectCount := 0
-
+		connectRetryLimit := 10
 		for {
```
[the remark this answers is not in the document]
Comment (thxCode) [reply pkg/agent/windows/remotedialer/client_windows.go]: updated

File: pkg/rkeworker/execute_linux.go (hunk 1 above)
Comment (thxCode) [reply pkg/rkeworker/execute_linux.go]: updated

File: pkg/rkeworker/execute_windows.go (hunk 2 above)
Comment (thxCode) [reply pkg/rkeworker/execute_windows.go]: updated

File: pkg/rkeworker/execute_windows.go (hunk 3 above)
Comment (thxCode) [reply pkg/rkeworker/execute_windows.go]: updated

File: pkg/rkeworker/execute_windows.go (hunk 4 above)
Comment (thxCode) [reply pkg/rkeworker/execute_windows.go]: I think here we only care about the `Retry` exist status and return other errors.

File: scripts/windows/build-agent.ps1 (hunk 5 above)
Comment (thxCode) [reply scripts/windows/build-agent.ps1]: `1809` host can package `1803` container with Hyper-V mode.

**Comment by orangedeng** [comment]
LGTM

**Comment by thxCode** [comment]
Fixing is including in https://github.com/rancher/rancher/pull/19717, close this.

**Pull request closed without merging**